File: benchmarks/java/B8JsonParse.java
import java.io.*;
import java.util.*;
// Simple recursive JSON stats counter using built-in tokenizer approach
public class B8JsonParse {
    static long[] count(Object v) {
        long[] r = new long[6]; // o,a,s,n,b,nl
        if (v instanceof Map) {
            r[0] = 1;
            for (Object val : ((Map<?,?>)v).values()) {
                long[] c = count(val);
                for (int i = 0; i < 6; i++) r[i] += c[i];
            }
        } else if (v instanceof List) {
            r[1] = 1;
            for (Object val : (List<?>)v) {
                long[] c = count(val);
                for (int i = 0; i < 6; i++) r[i] += c[i];
            }
        } else if (v instanceof String) r[2] = 1;
        else if (v instanceof Number) r[3] = 1;
        else if (v instanceof Boolean) r[4] = 1;
        else if (v == null) r[5] = 1;
        return r;
    }
    // Minimal JSON parser (no external deps)
    static int pos;
    static String json;
    static Object parse() {
        skipWS();
        char c = json.charAt(pos);
        if (c == '{') return parseObj();
        if (c == '[') return parseArr();
        if (c == '"') return parseStr();
        if (c == 't') { pos += 4; return Boolean.TRUE; }
        if (c == 'f') { pos += 5; return Boolean.FALSE; }
        if (c == 'n') { pos += 4; return null; }
        return parseNum();
    }
    static void skipWS() { while (pos < json.length() && " \t\n\r".indexOf(json.charAt(pos)) >= 0) pos++; }
    static Map<String,Object> parseObj() {
        Map<String,Object> m = new LinkedHashMap<>();
        pos++; skipWS();
        if (json.charAt(pos) == '}') { pos++; return m; }
        while (true) {
            skipWS(); String key = parseStr(); skipWS(); pos++; // :
            skipWS(); m.put(key, parse()); skipWS();
            if (json.charAt(pos) == '}') { pos++; return m; }
            pos++; // ,
        }
    }
    static List<Object> parseArr() {
        List<Object> a = new ArrayList<>();
        pos++; skipWS();
        if (json.charAt(pos) == ']') { pos++; return a; }
        while (true) {
            skipWS(); a.add(parse()); skipWS();
            if (json.charAt(pos) == ']') { pos++; return a; }
            pos++; // ,
        }
    }
    static String parseStr() {
        pos++; // opening "
        StringBuilder sb = new StringBuilder();
        while (json.charAt(pos) != '"') {
            if (json.charAt(pos) == '\\') { pos++; sb.append(json.charAt(pos)); }
            else sb.append(json.charAt(pos));
            pos++;
        }
        pos++; // closing "
        return sb.toString();
    }
    static Number parseNum() {
        int start = pos;
        while (pos < json.length() && "0123456789.eE+-".indexOf(json.charAt(pos)) >= 0) pos++;
        String s = json.substring(start, pos);
        if (s.contains(".") || s.contains("e") || s.contains("E")) return Double.parseDouble(s);
        return Long.parseLong(s);
    }
    public static void main(String[] args) throws Exception {
        String fn = args.length > 0 ? args[0] : "../../data/json_input.json";
        json = new String(java.nio.file.Files.readAllBytes(java.nio.file.Path.of(fn)));
        pos = 0;
        Object v = parse();
        long[] r = count(v);
        System.out.printf("objects=%d arrays=%d strings=%d numbers=%d bools=%d nulls=%d%n", r[0],r[1],r[2],r[3],r[4],r[5]);
    }
}

File: benchmarks/kotlin/b8_json_parse.kt
// Use same hand-rolled parser approach as Java
import java.io.File
var pos = 0
lateinit var json: String
fun skipWS() { while (pos < json.length && json[pos] in " \t\n\r") pos++ }
fun parse(): Any? {
    skipWS(); val c = json[pos]
    return when {
        c == '{' -> parseObj()
        c == '[' -> parseArr()
        c == '"' -> parseStr()
        c == 't' -> { pos += 4; true }
        c == 'f' -> { pos += 5; false }
        c == 'n' -> { pos += 4; null }
        else -> parseNum()
    }
}
fun parseObj(): Map<String, Any?> {
    val m = mutableMapOf<String, Any?>(); pos++; skipWS()
    if (json[pos] == '}') { pos++; return m }
    while (true) {
        skipWS(); val k = parseStr(); skipWS(); pos++ // :
        skipWS(); m[k] = parse(); skipWS()
        if (json[pos] == '}') { pos++; return m }; pos++ // ,
    }
}
fun parseArr(): List<Any?> {
    val a = mutableListOf<Any?>(); pos++; skipWS()
    if (json[pos] == ']') { pos++; return a }
    while (true) {
        skipWS(); a.add(parse()); skipWS()
        if (json[pos] == ']') { pos++; return a }; pos++ // ,
    }
}
fun parseStr(): String {
    pos++; val sb = StringBuilder()
    while (json[pos] != '"') {
        if (json[pos] == '\\') { pos++; sb.append(json[pos]) } else sb.append(json[pos]); pos++
    }; pos++; return sb.toString()
}
fun parseNum(): Number {
    val start = pos
    while (pos < json.length && json[pos] in "0123456789.eE+-") pos++
    val s = json.substring(start, pos)
    return if ('.' in s || 'e' in s || 'E' in s) s.toDouble() else s.toLong()
}
fun count(v: Any?): LongArray {
    val r = LongArray(6)
    when (v) {
        is Map<*, *> -> { r[0] = 1; v.values.forEach { c -> val x = count(c); for (i in 0..5) r[i] += x[i] } }
        is List<*> -> { r[1] = 1; v.forEach { c -> val x = count(c); for (i in 0..5) r[i] += x[i] } }
        is String -> r[2] = 1
        is Number -> r[3] = 1
        is Boolean -> r[4] = 1
        null -> r[5] = 1
    }; return r
}
fun main(args: Array<String>) {
    val fn = if (args.isNotEmpty()) args[0] else "../../data/json_input.json"
    json = File(fn).readText(); pos = 0
    val v = parse(); val r = count(v)
    println("objects=${r[0]} arrays=${r[1]} strings=${r[2]} numbers=${r[3]} bools=${r[4]} nulls=${r[5]}")
}
